### PR TITLE
Add `--force` flag to cargo install

### LINKED ISSUE #2082: cargo install --reinstall?
Currently `cargo install` won't overwrite existing package if asked to install one. However, this is very inconvenient when you're testing your own binaries (after each change you need to execute two commands instead of one to get a fresh version in PATH) or if you need a binary which frequently changes upstream.

What do you think of adding a `--reinstall` key to `cargo install` command? When used, it would be equivalent to `cargo uninstall <args> && cargo install <args without --reinstall>`, or something like it. I remember it was discussed in the RFC on `cargo install`, but I didn't manage to participate in that discussion; maybe it is worth to lay out reasons why it wasn't added eventually and probably reconsider that decision, if there was one.

### PR DESCRIPTION
**Pull request by gkoz**
Close #2082.

Adding `--force` (`-f`) instructs cargo to overwrite existing binaries (updating the metadata accordingly).
This allows updating crates via `cargo install -f <crate>`.

Installation happens in two stages now: binaries are copied into a temporary subdirectory of the destination first, then moved into destination. This should catch some errors earlier.

In case of installation error cargo will remove new binaries but won't attempt to undo successful overwrites.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR @gkoz! There's a few technical aspects I'd tweak here, but I agree that the first thing to do here is probably to decide whether we want to do this in the first place. Let's discuss more on the issue, though, as that's probably the best place to have this discussion.

(I'll write a comment over there shortly)

File: src/cargo/ops/cargo_install.rs [hunk 1]
```diff
@@ -190,18 +201,23 @@ fn one<I, F>(mut i: I, f: F) -> CargoResult<Option<I::Item>>
...
         }
         let mut msg = format!("binary `{}` already exists in destination", name);
         if let Some((p, _)) = prev.v1.iter().find(|&(_, v)| v.contains(&name)) {
-            msg.push_str(&format!(" as part of `{}`", p));
+            if replace {
+                return Ok(Some(p.clone()))
+            }
+            msg.push_str(&format!(" as part of `{}` (use --replace to override)", p));
         }
         Err(human(msg))
```
Comment (gkoz) [inline src/cargo/ops/cargo_install.rs]: I've realized that it's safer to always list all conflicts when run without `--replace` thus the error should probably be generated somewhere up the stack.

**Comment by gkoz** [comment]
There doesn't seem to be any opposition to this change.

**Comment by alexcrichton** [comment]
Ok, @wycats and I had a chance to talk about this recently, and this seems like it's desirable as it's pretty orthogonal from the `--update` flag. Put another way, `cargo install --replace` is distinct from `cargo install --update --replace`, so it seems fine to have both!

One thought we had was perhaps this could be renamed `-f` or `--force`? That follows the convention of many other tools for saying "just do this". Also, from the discussion on #2082 it sounds like the opinions about `--update` may have changed in light of our current situation, so if you'd like to implement that as well here feel free!

**Comment by gkoz** [comment]
Yes `--force` would work. I had avoided it to sidestep a bigger question of whether `cargo` should have such a flag and consistency considerations (if `install` gets it other commands might want to as well).

I'll keep `--update` in mind but wouldn't do it straight away.

You had mentioned some implementation issues. I'm guessing you don't like the `replace` (or `force`) flag being passed as a `bool` from `bin/install.rs`?

**Comment by alexcrichton** [comment]
Ah I believe what I had in mind was the "transactional" nature of the install. Right now the uninstall happens first, but it probably wants to get delayed until right before the actual new binaries are moved into place (either that or nothing is really uninstalled, just metadata is updated).

**Comment by alexcrichton** [comment]
ping @gkoz, any update on this?

**Comment by gkoz** [comment]
Oh wow, it's been a month? I'll revisit this in a few days.

**Comment by alexcrichton** [comment]
Sounds good to me! Feel free to ping me if you need help pushing over the finish line.

**Comment by gkoz** [comment]
> Ah I believe what I had in mind was the "transactional" nature of the install. Right now the uninstall happens first, but it probably wants to get delayed until right before the actual new binaries are moved into place (either that or nothing is really uninstalled, just metadata is updated).

To avoid any misunderstanding... The current patch calls `uninstall` after building but before copying the binaries. If that is too risky (after uninstalling we might find there's not enough space for the new binaries?) I could perhaps create a temporary subdirectory (`TempDir::new_in(&dst)`), copy new binaries into it and then overwrite the old ones.

**Comment by alexcrichton** [comment]
Basically what I'm getting at is two properties:

1. If a normal error happens for whatever reason, we shouldn't tamper with the install root at all
2. If ctrl-c happens, the "window of where things are corrupted" should be as small as possible

So given that, the error that I'm worried about here is that we successfully remove something but then hit an I/O error copying in a new file, In that case we won't restore the just-removed binaries. Could this be updated to handle that?

**Comment by gkoz** [comment]
I've pushed a version, that copies all binaries into a temp subdirectory of the destination and only after that succeeds tries to replace the existing binaries. If overwriting is partially successful, it doesn't try to undo anything (should it? that could further increase complexity) but still updates the metadata.

Another difference from former behavior is (when `--force` is absent) listing all of the binaries to be overwritten, not just the first one. However the error message still doesn't make it very clear that without `--bin` the old packages will be removed completely.

Error handling is not ideal here:
- I'm not sure how to make several lines of "binary ... already exists ..." and the "Add --force to overwrite" hint look nicer,
- there doesn't seem to be a way to report two unrelated but equally important errors.

Oh and I haven't tested on Windows yet to make sure there's no issue with `EXE_SUFFIX`.

File: src/cargo/ops/cargo_install.rs [hunk 2]
```diff
@@ -114,14 +107,90 @@ pub fn install(root: Option<&str>,
...
+                                       .collect::<HashSet<_>>();
+        // Remove stale records of overwritten binaries.
+        for bin in installed_names.iter() {
+            if let Some(&Some(ref p)) = duplicates.get(bin) {
+                if let Some(set) = list.v1.get_mut(p) {
+                    set.remove(bin);
+                }
+            }
+        }
+        // If `--bin` or `--example` weren't specified, uninstall obsolete
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_install.rs]: This seems a little surprising to me, this is basically saying if package A installed foo and bar, then B installed foo, you'd remove A's bar? I'd probably expect that A's bar stuck around and you could uninstall it with `cargo uninstall A`

File: src/cargo/ops/cargo_install.rs [hunk 3]
```diff
@@ -27,23 +28,12 @@ struct CrateListingV1 {
     v1: BTreeMap<PackageId, BTreeSet<String>>,
 }
 
-struct Transaction {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_install.rs]: I think we may still want this transaction type in some form. Before we write out the metadata of *new* binaries if an error happens we need to remove them. If we overwrite something an then an error happens, it's fine to throw our hands up in the air

**Comment by alexcrichton** [comment]
Thanks @gkoz! Can you also write some tests for some of the corner cases here? Just to make sure the metadata is managed properly and everything works as expected.

File: src/cargo/ops/cargo_install.rs (hunk 2 above)
Comment (gkoz) [inline src/cargo/ops/cargo_install.rs]: All this time I'd been assuming it's a good idea for `cargo install -f foo` to completely remove the remnants of the old package by default because that's probably what the user wants. :) If that's not the case I should probably add a hint "run 'cargo uninstall foo:{previous version}' to remove remaining binaries."

Comment (alexcrichton) [inline src/cargo/ops/cargo_install.rs]: Hm yeah I guess it depends on how you look at it. On one hand you could have installed a package with a ton of binaries, some of which you never used, and then you overwrite one of the never used ones. On the other hand this could also hamstring a package to be useless if it's the one binary that counts.

To me though `-f` just means "do this anyway" which doesn't necessarily imply "do the extra work of removing everything else", and I like the idea of adding a note that other binaries from the package are lying around and may wish to be removed.

**Comment by gkoz** [comment]
Still working on adding tests. Wondering how to imitate failure modes.

**Comment by alexcrichton** [comment]
It's ok to not necessarily exhaustively test them, it's fine to just smoke test a few simple ones like something already exists, a build failure when overwriting, etc.

**Comment by gkoz** [comment]
Added more tests and an optional optimization in a separate commit.

**Comment by alexcrichton** [comment]
@bors: r+ 9f0fa249b5d307bd97c8c4957a98ad9f47116bf9

Thanks!
